### PR TITLE
JBPM-7230 - Case instance migration

### PR DESCRIPTION
**Pull request by mswiderski**
depends on https://github.com/kiegroup/jbpm/pull/1218

### CODE REVIEW

**Comment by mswiderski** [comment]
Jenkins retest this

File: kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        Map<String, Object> mapping = Collections.emptyMap();
+        if (payload != null) {
+            logger.debug("About to unmarshal mapping from payload: '{}' using container {} marshaller", payload, containerId);
+            mapping = marshallerHelper.unmarshal(containerId, payload, marshallingType, Map.class);
+        }
+        if (mapping == null || !mapping.containsKey("ProcessMapping")) {
+            throw new IllegalArgumentException("Case instance migration requires ProcessMapping to be provided");
+        }
+        Map<String, String> processMapping = (Map<String, String>) mapping.getOrDefault("ProcessMapping", Collections.emptyMap());
+        Map<String, String> nodeMapping = (Map<String, String>) mapping.getOrDefault("ProcessMapping", Collections.emptyMap());
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java]: there should be "NodeMapping" I guess.

File: kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public CaseMigrationReportInstance migrateCaseInstance(String containerId, String caseId, String targetContainerId, String payload, String marshallingType) {
+        Map<String, Object> mapping = Collections.emptyMap();
+        if (payload != null) {
+            logger.debug("About to unmarshal mapping from payload: '{}' using container {} marshaller", payload, containerId);
+            mapping = marshallerHelper.unmarshal(containerId, payload, marshallingType, Map.class);
+        }
+        if (mapping == null || !mapping.containsKey("ProcessMapping")) {
+            throw new IllegalArgumentException("Case instance migration requires ProcessMapping to be provided");
+        }
+        Map<String, String> processMapping = (Map<String, String>) mapping.getOrDefault("ProcessMapping", Collections.emptyMap());
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java]: I think we already know that "ProcessMapping" key is present so getOrDefault is probably not needed.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseInstanceMigrationIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        processInstances = caseClient.getProcessInstances(CONTAINER_ID_2, caseId, Arrays.asList(1), 0, 10);
+        assertThat(processInstances).isNotNull();
+        assertThat(processInstances).hasSize(2);
+        
+        for (ProcessInstance pi : processInstances) {
+            assertThat(pi.getContainerId()).isEqualTo(CONTAINER_ID_2);   
+        }
+    }
+
+
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseInstanceMigrationIntegrationTest.java]: Does it make sense to include nodeMapping test and one error test?

File: kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java (hunk 2 above)
Comment (mswiderski) [reply kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java]: good point, replaced with simple get

File: kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java (hunk 1 above)
Comment (mswiderski) [reply kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseAdminServiceBase.java]: very good catch :) fixed

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-case-mgmt/src/build/revapi-config.json [hunk 4]
```diff
@@ -201,6 +201,45 @@
...
+          "new": "class org.kie.server.remote.rest.casemgmt.CaseAdminResource",
+          "annotationType": "javax.ws.rs.Path",
+          "annotation": "@javax.ws.rs.Path(\"server\/admin\")",
+          "attribute": "value",
+          "oldValue": "\"server\/admin\/cases\"",
+          "newValue": "\"server\/admin\"",
+          "package": "org.kie.server.remote.rest.casemgmt",
+          "classSimpleName": "CaseAdminResource",
+          "elementKind": "class",
+          "justification": "Added support for case instance migration"
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-case-mgmt/src/build/revapi-config.json]: the change was needed to allow more fine grained method annotations so it was just a refactor

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseInstanceMigrationIntegrationTest.java (hunk 3 above)
Comment (mswiderski) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseInstanceMigrationIntegrationTest.java]: be my guest ;) in other words would be good that someone else tries the api to spot any issues

Comment (MarianMacik) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseInstanceMigrationIntegrationTest.java]: that's exactly what I thought about :) Will include these test perhaps some time this or next week.

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Pull request merged**